File: 명찬/W1/체육복.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    int answer = 0;
    vector <int> v(n + 2, 1);
    for (int i = 0; i < lost.size(); i++)
    {
        v[lost[i]] = 0;
    }
    for (int i = 0; i < reserve.size(); i++)
    {
        v[reserve[i]] += 1;
    }

    for (int j = 0; j < n / 2; j++) //전달후 다시 검토 n/2번 수행.
    {
        for (int i = 1; i < n + 1; i++) //한쪽에만 도난당한 친구가 있으면 전달.
        {
            if (v[i] == 2)
            {
                if (v[i - 1] == 0 && v[i + 1] != 0)
                {
                    v[i - 1] = 1;
                    v[i] = 1;
                }
                if (v[i - 1] != 0 && v[i + 1] == 0)
                {
                    v[i + 1] = 1;
                    v[i] = 1;
                }
            }
        }
    }
    for (int i = 1; i < n + 1; i++) //이후 양쪽에 둘다 체육복이 없는 경우와 있는 경우 처리
    {

        if (v[i] == 1)
        {
            answer++;
        }
        else if (v[i] == 2 && v[i - 1] == 0 && v[i + 1] == 0) //양쪽에 있으면 한쪽에 전달.
        {
            answer += 2;
        }
        else if (v[i] == 2)
        {
            answer++;
        }
    }
    return answer;
}

File: 채현/W1/체육복.py
# 처음에는 단순히 lost 학생이 앞번호(i-1), 없으면 뒷번호(i+1)에게 체육복을 빌리는 방식으로 생각했다.
# 그런데 전체 테스트에서는 통과하지 못하는 케이스가 있었다.

# 문제를 다시 보니 여벌이 있는 학생도 도난당할 수 있어서,
# lost와 reserve에 동시에 포함된 학생을 먼저 처리하도록 수정했다.
# 하지만 이 경우에도 일부 케이스가 여전히 통과하지 않았다.

# 반복문으로 lost를 순회하면서 동시에 lost.remove()를 호출하고 있었는데
# 이렇게 되면 리스트가 순회 도중 변경되어 일부 원소를 건너뛸 수 있다.

# 그래서 lost를 직접 수정하지 않고, 실제로 체육복이 없는 학생만 new_lost에 따로 담아 처리하도록 바꿨다.

def solution(n, lost, reserve):
    lost.sort()
    reserve.sort()
            
    new_lost = []
    for x in lost:
        if x in reserve:
            reserve.remove(x)
        else:
            new_lost.append(x)

    lost = new_lost
    
    for i in lost:
        if (i - 1) in reserve:
            reserve.remove(i - 1)
        elif (i + 1) in reserve:
            reserve.remove(i + 1)
        else:
            n -= 1
            
    answer = n
    return answer
